python/lsst/daf/butler/datastores/posixDatastore.py

```diff
@@ -112,127 +186,200 @@ def get(self, uri, storageClass, parameters=None):
+        if not isinstance(inMemoryDataset, storageClass.pytype):
+            raise ValueError("Inconsistency between supplied object ({}) "
+                             "and storage class type ({})".format(type(inMemoryDataset), storageClass.pytype))
 
+        # Work out output file name
         template = self.templates.getTemplate(typeName)
-        location = self.locationFactory.fromPath(template.format(dataUnits,
+        location = self.locationFactory.fromPath(template.format(ref.dataId,
                                                                  datasetType=typeName))
 
-        # Write a single component
-        formatter = self.formatterFactory.getFormatter(storageClass, typeName)
+        # Get the formatter based on the storage class
+        formatter = self.formatterFactory.getFormatter(datasetType.storageClass, typeName)
 
         storageDir = os.path.dirname(location.path)
         if not os.path.isdir(storageDir):
             safeMakeDir(storageDir)
-        return formatter.write(inMemoryDataset, FileDescriptor(location, pytype=storageClass.pytype,
+
+        # Write the file
+        path = formatter.write(inMemoryDataset, FileDescriptor(location,
                                                                storageClass=storageClass))
 
-    def remove(self, uri):
-        """Indicate to the Datastore that a `Dataset` can be removed.
+        # Create Storage information in the registry
+        ospath = os.path.join(self.root, path)
+        checksum = self.computeChecksum(ospath)
+        stat = os.stat(ospath)
+        size = stat.st_size
+        info = StorageInfo(self.name, checksum, size)
+        self.registry.addStorageInfo(ref, info)
+
+        # Associate this dataset with the formatter for later read.
+        fileInfo = StoredFileInfo(formatter, path, storageClass)
+        self.addStoredFileInfo(ref, fileInfo)
+
+        # Register all components with same information
+        for compRef in ref.components.values():
+            self.registry.addStorageInfo(compRef, info)
+            self.addStoredFileInfo(compRef, fileInfo)
+
+    def getUri(self, ref):
+        """URI to the Dataset.
 
         Parameters
         ----------
+        ref : `DatasetRef`
+            Reference to the required Dataset.
+
+        Returns
+        -------
         uri : `str`
-            A Universal Resource Identifier that specifies the location of the
-            stored `Dataset`.
+            URI string pointing to the Dataset within the datastore. If the
+            Dataset does not exist in the datastore, the URI may be a guess
+            and will include a URI fragment "#predicted".
+            If the datastore does not have entities that relate well
+            to the concept of a URI the returned URI string will be
+            descriptive. The returned URI is not guaranteed to be obtainable.
+
+        """
+
+        # if this has never been written then we have to guess
+        if ref.id is None:
```

Comment:
This check only tells you whether *some* `Datastore` has written it, not whether this `Datastore` has, and I think that's what you want to be checking.

That said, I think users actually may want to specify (e.g. with a boolean arg) whether they're asking for a guess URI for something they haven't written yet or a URI for something that should exist, so they can get an exception when they don't get what they want.

Reply:
When I wrote this I didn't have an exists method. I'll add an option to only retrieve things that actually exist.